src/wtf8/mod.rs

```diff
@@ -388,9 +462,16 @@ impl<T: Wtf8Index> Index<T> for Wtf8 {
 impl AsRef<Wtf8> for str {
     #[inline]
     fn as_ref(&self) -> &Wtf8 {
-        // Safety: the cast is sound because repr(transparent), matching the layout of str.
-        // UTF-8 is a subset of WTF-8, so type invariants are never violated.
-        unsafe { &*(self as *const str as *const Wtf8) }
+        // Safety: UTF-8 is a subset of WTF-8.
+        unsafe { Wtf8::from_bytes_unchecked(self.as_bytes()) }
+    }
+}
+
+impl AsMut<Wtf8> for str {
+    #[inline]
+    fn as_mut(&mut self) -> &mut Wtf8 {
+        // Safety: UTF-8 is a subset of WTF-8.
+        unsafe { Wtf8::from_bytes_unchecked_mut(self.as_bytes_mut()) }
     }
 }

```

Comment:
```suggestion
```

This is unsound for the same reason `bytes_mut` is unsafe.

One can cast a reference of `str` to `Wtf8`, modify it so that it is valid `Wtf8` but not UTF-8, and then the result is undefined behaviour.

For example:

```rust
let mut utf8 = String::from("abcdef");
let mut wtf8 = Wtf8::new_mut(utf8.as_mut_str());
modify_somehow(&mut wtf8);
print!("{}", utf8);
```

This would compile and produce undefined behaviour if `modify_somehow` introduces unpaired surrogates to the string.